### ISSUE TITLE
error: io.EOF

### ISSUE THREAD

**Issue by Impyy** [vmihailenco/msgpack#113]
The following code causes an io.EOF error if the amount of Objects is larger than 1.
```go
package main

import "gopkg.in/vmihailenco/msgpack.v2"

type innerObj struct {
	Objects []*obj
}

type obj struct {
	inner *innerObj
}

func main() {
	o1 := obj{
		// works:
		// inner: &innerObj{[]*obj{&obj{}}},

		// doesn't work:
		inner: &innerObj{[]*obj{&obj{}, &obj{}}},
	}

	bytes, err := msgpack.Marshal(&o1)
	if err != nil {
		panic(err)
	}

	o2 := obj{}
	if err = msgpack.Unmarshal(bytes, &o2); err != nil {
		panic(err)
	}
}

func (o *obj) MarshalMsgpack() ([]byte, error) {
	return msgpack.Marshal(o.inner)
}

func (o *obj) UnmarshalMsgpack(data []byte) error {
	inner := new(innerObj)
	if err := msgpack.Unmarshal(data, inner); err != nil {
		return err
	}

	o.inner = inner
	return nil
}
```
For the record, a similar construction does work with Go's xml package.

**Comment by vmihailenco** [comment]
AFAIK this is limitation of msgpack:
- you have slice with 2 nil elements encodes as: `92 c0 c0`
- lib reads 92 - slice with 2 elements
- now it has to decode `c0 c0` so it passes it to `UnmarshalMsgpack([]byte{c0, c0})`
- and at this point we don't know how many bytes `UnmarshalMsgpack` has processed decoding first object so we stop with `io.EOF`

That is why I added `CustomEncoder` and `CustomDecoder`:

```
package main

import (
	"encoding/hex"
	"log"

	"gopkg.in/vmihailenco/msgpack.v2"
)

type innerObj struct {
	Objects []*obj
}

type obj struct {
	inner *innerObj
}

func main() {
	o1 := obj{
		// works:
		// inner: &innerObj{[]*obj{&obj{}}},

		// doesn't work:
		inner: &innerObj{[]*obj{&obj{}, &obj{}}},
	}

	bytes, err := msgpack.Marshal(&o1)
	if err != nil {
		panic(err)
	}

	log.Println(hex.Dump(bytes))

	o2 := obj{}
	if err = msgpack.Unmarshal(bytes, &o2); err != nil {
		panic(err)
	}
}

func (o *obj) EncodeMsgpack(e *msgpack.Encoder) error {
	return e.Encode(o.inner)
}

func (o *obj) DecodeMsgpack(d *msgpack.Decoder) error {
	inner := new(innerObj)
	if err := d.Decode(inner); err != nil {
		return err
	}

	o.inner = inner
	return nil
}
```

**Comment by vmihailenco** [comment]
Actually you example can be fixed since you encode only 1 element. Fixing it :)

**Comment by vmihailenco** [comment]
Fixed in v2.8.2, but note that my example is slightly faster and more flexible.

**Comment by Impyy** [comment]
Great, it works. Thanks for the quick fix!
